### ISSUE TITLE
Parallelize garbage collection to prevent snowballing of volumes/containers

### ISSUE THREAD

**Issue by vito** [concourse/concourse#1297]
Today's GC is very naive: we garbage collect everything serially, in order. Granted, we've optimized the order to catch most things in a single pass, but that's about the only good thing about it.

On our large Wings deployment we observed a very slow-to-recover (possibly never?) situation wherein the ATC was taking so long to garbage collect volumes that the number of containers rose. So the next GC pass had to garbage collect even more containers, meanwhile the number of volumes rose. This see-sawing happened and kept escalating until I intervened and kept restarting the ATC midway through its volume/container collection so it could switch to the other.

The most trivial start to this is to GC containers and volumes in parallel, i.e. just put a `go` around them. They'll still be GCed serially, so it's not perfect, but it should at least fix the see-sawing.

Long-term, I think we'll want something like a worker pool/queue model per-worker so we can parallelize the individual container/volume collection without overwhelming the workers.

**Comment by vito** [comment]
Update: we did the initial easy step of GCing volumes + containers in parallel (still synchronous for each), but even with that we still ran in to a case where the containers crept up, perhaps because of a slow/unresponsive worker.

**Comment by vito** [comment]
Here are some vectors to consider for making garbage collection not prone to slowing itself down:

* [ ] Having many containers.
* [ ] Having many volumes.
* [ ] A single slow worker (overall).
* [ ] A single stuck container delete on the worker.
* [ ] A single stuck volume delete on the worker.
* [ ] A single stuck call to the database (e.g. mark container/volume as destroying).

**Comment by mhuangpivotal** [comment]
Reopening because we are consistently seeing our number of containers go up in ci ever since these commits. We are leaking containers somehow.

**Comment by jtarchie** [comment]
We saw these yesterday while deploying. We think they might be symptoms. 

![screen shot 2017-07-24 at 2 42 08 pm](https://user-images.githubusercontent.com/75184/28574902-86aa0a9a-710c-11e7-9e97-e47fcc80a5dd.png)
![screen shot 2017-07-24 at 2 42 16 pm](https://user-images.githubusercontent.com/75184/28574900-86a93444-710c-11e7-8222-a0b50dc2f288.png)
![screen shot 2017-07-24 at 2 43 11 pm](https://user-images.githubusercontent.com/75184/28574903-86ae2c10-710c-11e7-97a1-4b096d056984.png)
<img width="641" alt="screen shot 2017-07-24 at 2 43 28 pm" src="https://user-images.githubusercontent.com/75184/28574899-86a8e714-710c-11e7-9d31-1c815bc32cc0.png">
![screen shot 2017-07-24 at 8 20 35 am](https://user-images.githubusercontent.com/75184/28574901-86a91fb8-710c-11e7-8c3f-7488d18729ca.png)
![screen shot 2017-07-24 at 8 32 00 am](https://user-images.githubusercontent.com/75184/28574904-86b13f22-710c-11e7-9933-3b80cb6e500d.png)

**Comment by topherbullock** [comment]
We're observing occasional issues where radar's resource checking containers are swept out from under its feet during resource checking, and the resource errors showing the following message: 

`open /proc/123456/ns/net: no such file or directory`

Relevant papertrail logs show the following:

```
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.202791691","source":"guardian","message":"guardian.destroy.destroy.delete.started","log_level":1,"data":{"handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132150.1.2"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.327237606","source":"guardian","message":"guardian.create.network.finished","log_level":1,"data":{"handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132148.5","spec":""}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.327345848","source":"guardian","message":"guardian.create.create-failed-cleaningup.start","log_level":1,"data":{"cause":"open /proc/2917915/ns/net: no such file or directory","handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132148.6"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.327390909","source":"guardian","message":"guardian.create.create-failed-cleaningup.destroy.started","log_level":1,"data":{"cause":"open /proc/2917915/ns/net: no such file or directory","handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132148.6.1"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.355417013","source":"guardian","message":"guardian.destroy.destroy.delete.finished","log_level":1,"data":{"handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132150.1.2"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.355458260","source":"guardian","message":"guardian.destroy.destroy.finished","log_level":1,"data":{"handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132150.1"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.387221575","source":"guardian","message":"guardian.create.create-failed-cleaningup.destroy.state-failed-skipping-delete","log_level":1,"data":{"cause":"open /proc/2917915/ns/net: no such file or directory","error":"runc state: runc create: exit status 1: container \"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5\" does not exist\n","handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132148.6.1"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.387291431","source":"guardian","message":"guardian.create.create-failed-cleaningup.destroy.finished","log_level":1,"data":{"cause":"open /proc/2917915/ns/net: no such file or directory","handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132148.6.1"}}
Jul 25 06:40:39 worker-0 garden:  {"timestamp":"1500990039.388018131","source":"guardian","message":"guardian.create.containerizer-create.watch.done","log_level":1,"data":{"handle":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","session":"132148.2.4"}}
Jul 25 06:40:39 web-0 atc:  {"timestamp":"1500990038.889587402","source":"atc","message":"atc.syncer.images:radar.scan-resource.interval-runner.tick.created-creating-container-in-db","log_level":0,"data":{"container":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","pipeline-scoped-name":"images:resource:concourse-topgun-ci-image","session":"18.5.196.1.174"}}
Jul 25 06:40:39 web-0 atc:  {"timestamp":"1500990038.890908241","source":"atc","message":"atc.syncer.images:radar.scan-resource.interval-runner.tick.acquire.acquired","log_level":0,"data":{"container":"1b1718b5-6ac2-48ac-64e5-cef7b3eaded5","id":[5,3130700],"pipeline-scoped-name":"images:resource:concourse-topgun-ci-image","session":"18.5.196.1.174.1"}}
```

**Comment by topherbullock** [comment]
Took a crack at writing an integration test that reproduces this interaction between radar and container collector... still WIP.
https://github.com/concourse/atc/tree/wip-1297

**Comment by mhuangpivotal** [comment]
Looking at the Papertrail logs for the past month, the first instance of the above error occurred at `2017-06-30T07:28:20 UTC`.

**Comment by topherbullock** [comment]
Made a separate issue for the `open /proc/123456/ns/net: no such file or directory` issue, which captures the GC -> Radar behaviour more succinctly.

https://github.com/concourse/concourse/issues/1413
